File: v1.70.0/abs/src/bin/abc049c.rs
use proconio::input;
use proconio::marker::Chars;

fn main() {
    input! {
        mut s: Chars,
    }
    s.reverse();

    let dreamer = "dreamer".chars().rev().collect::<String>();
    let eraser = "eraser".chars().rev().collect::<String>();
    let dream = "dream".chars().rev().collect::<String>();
    let erase = "erase".chars().rev().collect::<String>();

    let mut i = 0;
    while i < s.len() {
        let cur7 = if i + 6 < s.len() {
            Some(s.iter().skip(i).take(7).collect::<String>())
        } else {
            None
        };

        let cur6 = if i + 5 < s.len() {
            Some(s.iter().skip(i).take(6).collect::<String>())
        } else {
            None
        };

        let cur5 = if i + 4 < s.len() {
            Some(s.iter().skip(i).take(5).collect::<String>())
        } else {
            None
        };

        if cur7 == Some(dreamer.clone()) {
            i += 7;
        } else if cur6 == Some(eraser.clone()) {
            i += 6;
        } else if cur5 == Some(dream.clone()) || cur5 == Some(erase.clone()) {
            i += 5;
        } else {
            println!("NO");
            return;
        }
    }

    println!("YES");
}

File: csharp/7.0.7/abs/abc049c.cs
using System;
using System.Linq;

class Program
{
    static void Main(string[] args)
    {
        char[] s = Console.ReadLine().ToCharArray();
        Array.Reverse(s);
        int i = 0;
        while (i < s.Length)
        {
            // Console.WriteLine(i);
            if (i + 6 < s.Length && new string(s, i, 7) == "remaerd")
            {
                // dreamer
                i += 7;
            }
            else if (i + 4 < s.Length && new string(s, i, 5) == "maerd")
            {
                // dream
                i += 5;
            }
            else if (i + 5 < s.Length && new string(s, i, 6) == "resare")
            {
                // eraser
                i += 6;
            }
            else if (i + 4 < s.Length && new string(s, i, 5) == "esare")
            {
                // erase
                i += 5;
            }
            else
            {
                Console.WriteLine("NO");
                return;
            }
        }

        Console.WriteLine("YES");
    }
}
